### PR TITLE
Allow discriminant reads not followed by `SwitchInt`

### LINKED ISSUE #120: Support `Rvalue::Discriminant` not followed by `SwitchInt`
This can show up in optimized MIR, for example the optimized mir of `Option::is_some()`. It's not too hard to support: we can replace `_x = Discriminant(_y)` with
```rust
match _y {
    Variant1 => _x = <discriminant value for variant 1>,
    ...
}
```

### PR DESCRIPTION
**Pull request by Nadrieril**
We replace `_x = Discr(_y);` not followed by a `SwitchInt` with
```rust
match _y {
    0 => {
        _x := const (0 : isize)
    },
    1 => {
        _x := const (1 : isize)
    }
    ...
}
```

Fixes #120.

### CODE REVIEW

**Comment by Nadrieril** [comment]
We can use this to make https://github.com/AeneasVerif/charon/issues/91 work too, as soon as hax forwards the right information to us.

File: charon/tests/ui/issue-91-enum-to-discriminant-cast.out [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+fn test_crate::main()
+
+fn core::clone::Clone::clone<'_0, Self>(@1: &'_0 (Self)) -> Self
+
+
+
+error: aborting due to previous error
+
+[ ERROR charon_driver:264] The extraction encountered 1 errors
```
Comment (sonmarcho) [inline charon/tests/ui/issue-91-enum-to-discriminant-cast.out:58]: Why does this one fail? I thought this PR would fix it.

File: charon/src/values_utils.rs [hunk 2]
```diff
@@ -218,6 +218,9 @@ impl ScalarValue {
             ScalarValue::I128(v) => v as u128,
         }
     }
+    pub fn from_bits(ty: IntegerTy, bits: u128) -> Self {
```
Comment (sonmarcho) [inline charon/src/values_utils.rs:221]: Nit: could you add a space above?

File: charon/src/remove_read_discriminant.rs [hunk 3]
```diff
@@ -102,53 +69,106 @@ impl<'a, 'tcx, 'ctx> Visitor<'a, 'tcx, 'ctx> {
...
+                            };
+                            new_sequence(switch, st3).content
+                        } else {
+                            switch
+                        };
+                    }
+                    Err(st2) => {
+                        // The discriminant read is not followed by a `SwitchInt`. We replace `_x =
+                        // Discr(_y)` with `match _y { 0 => { _x = 0 }, 1 => { _x = 1; }, .. }`.
+                        let targets = variants
```
Comment (sonmarcho) [inline charon/src/remove_read_discriminant.rs:145]: Just add a comment which states that it happens when we do, for instance: `y as usize`.

**Review by sonmarcho — CHANGES_REQUESTED** [review]

File: charon/tests/ui/issue-91-enum-to-discriminant-cast.out (hunk 1 above)
Comment (Nadrieril) [reply charon/tests/ui/issue-91-enum-to-discriminant-cast.out:58]: `as` casts use a different MIR statement, and hax currently doesn't support it

File: charon/src/remove_read_discriminant.rs (hunk 3 above)
Comment (Nadrieril) [reply charon/src/remove_read_discriminant.rs:145]: That's not the case. As far as I know only optimizing a `match` can make that happen. I'll note that

File: charon/src/values_utils.rs (hunk 2 above)
Comment (Nadrieril) [reply charon/src/values_utils.rs:221]: FTR I like grouping together functions that belong together

Comment (sonmarcho) [reply charon/src/values_utils.rs:222]: I like having spaces, I think it's cleaner :)

File: charon/src/remove_read_discriminant.rs (hunk 3 above)
Comment (sonmarcho) [reply charon/src/remove_read_discriminant.rs]: Ok, then just add a comment which explains when this can happen :)

**Review by sonmarcho — APPROVED** [review]

**Comment by sonmarcho** [comment]
You're free to mark the conversation as resolved and merge when you consider they are.

**Pull request merged**